File: #202_Happy_Number/Solution.java
import java.util.*;

public class Solution {
    public boolean isHappy(int n) {
    	if (n == 1) return true;
    	if (n == 0) return false;
        Set<Integer> set = new HashSet<Integer>();
        while (n != 1) {
        	if (set.contains(n)) return false;
        	set.add(n);
        	int result = 0;
        	while (n != 0) {
        		int m = n;
        		n = n / 10;
        		int rem = m - n * 10;
        		result += rem * rem;
        	}
        	n = result;
        }
        return true;
    }

    public static void main(String[] args) {
    	Solution sol = new Solution();
    	System.out.println(sol.isHappy(19));
    }
}

File: #202_Happy_Number/Solution.py
class Solution(object):
    def isHappy(self, n):
    	computed = set()
    	while n != 1:
    		if n in computed:
    			return False
    		computed.add(n)
    		n = self.transform(n)
    	return True

    def transform(self, n):
    	re = 0
    	while n > 0:
    		[n, r] = divmod(n, 10)
    		re += r * r
    	return re

sol = Solution()
print sol.isHappy(20)